framework/contracts/native/version-control/src/commands.rs

```diff
@@ -31,32 +35,95 @@ use crate::{
-        return Err(VCError::NotAccountFactory {});
-    }
+    // Check that sender is account
+    let contract_info = cw2::query_contract_info(&deps.querier, &msg_info.sender)?;
+    let maybe_acc_module_info = ModuleInfo::try_from(contract_info)?;
+
+    ensure!(
+        maybe_acc_module_info.id() == ACCOUNT,
+        VCError::NotAccountInfo {
+            caller_info: maybe_acc_module_info
+        }
+    );
 
-    // Check if account already exists
+    // Ensure account isn't already registered
+    let account_id = ACCOUNT_ID.query(&deps.querier, msg_info.sender.clone())?;
     ensure!(
         !ACCOUNT_ADDRESSES.has(deps.storage, &account_id),
         VCError::AccountAlreadyExists(account_id)
     );
 
-    ACCOUNT_ADDRESSES.save(deps.storage, &account_id, &account_base)?;
+    let acc_module_info = maybe_acc_module_info;
+
+    let account_code_id = REGISTERED_MODULES
+        .load(deps.storage, &acc_module_info)?
+        .unwrap_account()?;
+
+    // provided and smaller, assert is eq to sequence
+    // provided and larger, just register
+    // Remote account_id is provided, assert the creator is the ibc host.
+    if account_id.is_local() {
+        // Predictable Account Id Sequence have to be >= 2147483648
+        if account_id.seq() < 2147483648 {
+            let next_sequence = LOCAL_ACCOUNT_SEQUENCE.may_load(deps.storage)?.unwrap_or(0);
+
+            ensure_eq!(
+                next_sequence,
+                account_id.seq(),
+                VCError::InvalidAccountSequence {
+                    expected: next_sequence,
+                    actual: account_id.seq(),
+                }
+            );
+
+            LOCAL_ACCOUNT_SEQUENCE.save(deps.storage, &next_sequence.checked_add(1).unwrap())?;
+        }
+    } else {
+        // If a non-local account_id is provided, assert that the creator is the ibc host
+        let creator_addr = deps.api.addr_validate(&creator)?;
+        let sender_cw2_info = cw2::query_contract_info(&deps.querier, &creator_addr)?;
+        let ibc_host_addr = REGISTERED_MODULES
+            .load(
+                deps.storage,
+                &ModuleInfo::from_id(IBC_HOST, sender_cw2_info.version.into())?,
+            )?
+            .unwrap_native()?;
+
+        ensure_eq!(
+            creator_addr,
+            ibc_host_addr,
+            VCError::SenderNotIbcHost(creator_addr.into_string(), ibc_host_addr.into())
+        );
+        // then assert that the account trace is remote and properly formatted
+        account_id.trace().verify_remote()?;
+    }
+
+    // verify code-id of sender
+    let sender_contract_info = deps.querier.query_wasm_contract_info(&msg_info.sender)?;
+
+    ensure_eq!(
+        account_code_id,
+        sender_contract_info.code_id,
+        VCError::NotAccountCodeId {
+            account_info: acc_module_info,
+            expected_code_id: account_code_id,
+            actual_code_id: sender_contract_info.code_id
+        }
+    );
```

Comment:
Let's move all validations to the start of the method

Reply:
done